### PR TITLE
CMake improvements

### PR DESCRIPTION
**Pull request by friendlyanon**
This PR will modernize the CMake usage of this library.

Some things to note:
* For conan the `cmake` generator should **NOT** be used, you should be using at least the `cmake_find_package` to generate targets that can be discovered using `CMAKE_MODULE_PATH`. The more ideal generator would be `cmake_find_package_multi`, but conan v2 is not yet out.
* GTest was being compiled with every single CI run, instead now a script pulls in a known good commit of GTest, compiles it, installs it to a prefix location that can be discovered using `CMAKE_PREFIX_PATH` and caches GTest. While GTest does not support clients using test discovery, CMake does support clients using test discovery via the `GoogleTest` module, so `ctest` does not need `-VV` verbosity level to show a proper test progression summary.

To see how the above affects building the library, please see the CI workflows that have been adapted for this improved process.

This library will also provide a real relocatable CMake package for easy consumption by clients.

For the rationale behind the changes, please see my contributions to [FunctionalPlus](https://github.com/Dobiasd/FunctionalPlus) for related discussions.

### CODE REVIEW

**Comment by erikzenker** [comment]
Hi @friendlyanon thank you for your contribution. I will start to review your PR and might come up with some question. I think it will take me a while to go through all the changes.

**Comment by friendlyanon** [comment]
Resolved the conflicts by rebasing the changes on top of the target branch.

File: .github/workflows/windows-ci.yml [hunk 1]
```diff
@@ -2,30 +2,67 @@ name: Windows CI
...
+
+    - run: mkdir -p build/dependencies/conan
+
+    - name: Install Conan Dependencies
+      working-directory: build/dependencies/conan
+      shell: bash
+      env: { CONAN_USER_HOME: '${{ github.workspace }}/.conan' }
+      run: conan install ../../.. --build missing
+
+    - name: Cache GTest
```
Comment (erikzenker) [inline .github/workflows/windows-ci.yml:40]: Why not install gtest from conan now?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,42 +1,80 @@
...
+# Protect dependents from this project's warnings if the guard isn't disabled
+set(hsm_warning_guard SYSTEM)
+if(hsm_INCLUDE_WITHOUT_SYSTEM)
+  set(hsm_warning_guard "")
 endif()
 
-if(${HSM_TIME_TRACE})
-    set(CMAKE_CXX_FLAGS "-ftime-trace")
-endif()
+# ---- Declare library ----
```
Comment (erikzenker) [inline CMakeLists.txt:19]: Why was that moved from the `src` folder cmake here? I followed that template for my cmake https://github.com/erikzenker/cmake-project-template which is based on https://www.youtube.com/watch?v=S4QSKLXdTtA

**Review by erikzenker — COMMENTED** [review]
What is the rational behind moving from `src/include` to `include`?

File: .github/workflows/clang-tidy.yml [hunk 3]
```diff
@@ -1,26 +1,29 @@
...
-        clang-tidy-10 --version
+      - uses: actions/checkout@v2
+
+      - name: Install Dependencies
+        run: | 
+          sudo apt install clang-tidy-10 libboost-all-dev
+          clang-tidy-10 --version
+
+      - name: Clang tidy
+        run: >-
```
[the remark this answers is not in the document]
Comment (friendlyanon) [reply .github/workflows/clang-tidy.yml:24]: https://yaml-multiline.info/

File: .github/workflows/windows-ci.yml [hunk 4]
```diff
@@ -2,30 +2,67 @@ name: Windows CI
...
+
+    - name: Configure Hsm
+      run: >-
+        $pwd = (Get-Location) -replace "\\", "/"
+
+        & cmake -S test -B build/test
+        -D "CMAKE_MODULE_PATH=${pwd}/build/dependencies/conan"
+        -D "CMAKE_PREFIX_PATH=${pwd}/build/dependencies/gtest"
+
+        if ($LASTEXITCODE -ne 0) { throw "Exit code is $LASTEXITCODE" }
```
[the remark this answers is not in the document]
Comment (friendlyanon) [reply .github/workflows/windows-ci.yml:58]: The default shell in windows workflows is `pwsh`, so you need to do a little more to get the correct path format. This step makes sure the format is `C:/path/to/prefix` on Windows.

File: .github/workflows/windows-ci.yml (hunk 1 above)
Comment (friendlyanon) [reply .github/workflows/windows-ci.yml:40]: I don't really use conan myself, so I'm not aware of gtest existing on conan. The entirety of gtest was in this repo, so I thought it's just easier (for me) to do it directly.

File: CMakeLists.txt (hunk 2 above)
Comment (friendlyanon) [reply CMakeLists.txt:19]: It's a personal preference. When I look at the root CML, I expect to see **all** the build and usage requirements of the library, without having to chase many CML files.

**Comment by friendlyanon** [comment]
Moving the headers to `include` from `src/include` is the least surprising. When I see a `src` or `source` directory, it's reasonable to assume that a library builds artifacts, i.e. isn't header only. It's also more common to do it this way than not.

**Comment by erikzenker** [comment]
@friendlyanon thx again for your contribution. Out of curiosity: do you use the hsm somewhere?

**Comment by friendlyanon** [comment]
I don't use this library *yet*, but as you might have seen from my recent activity on GitHub, I am the change I want to see in the world. I just want projects to utilize CMake properly, so they are trivial to consume and package.
